File: src/src_core/fmg/core/img/BurgerMenuModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using fmg.common;
using fmg.common.geom;
using fmg.common.notyfier;

namespace fmg.core.img {

    /// <summary> MVC: model of representable menu as horizontal or vertical lines </summary>
    public sealed class BurgerMenuModel : IAnimatedModel {

        private AnimatedImageModel _generalModel;
        private bool _show = true;
        private bool _horizontal = true;
        private int _layers = 3;
        private BoundDouble? _padding;

        private bool _disposed;
        public event PropertyChangedEventHandler PropertyChanged;
        private readonly NotifyPropertyChanged _notifier;

        /// <summary> ctor </summary>
        /// <param name="generalModel">another basic model</param>
        internal BurgerMenuModel(AnimatedImageModel generalModel) {
            _notifier = new NotifyPropertyChanged(this, ev => PropertyChanged?.Invoke(this, ev));
            _generalModel = generalModel;
            _generalModel.PropertyChanged += OnGeneralModelPropertyChanged;
        }

        /// <summary> image width and height in pixel </summary>
        public SizeDouble Size          { get => _generalModel.Size         ; set => _generalModel.Size          = value; }
        public bool       Animated      { get => _generalModel.Animated     ; set => _generalModel.Animated      = value; }
        public long       AnimatePeriod { get => _generalModel.AnimatePeriod; set => _generalModel.AnimatePeriod = value; }
        public int        TotalFrames   { get => _generalModel.TotalFrames  ; set => _generalModel.TotalFrames   = value; }
        public int        CurrentFrame  { get => _generalModel.CurrentFrame ; set => _generalModel.CurrentFrame  = value; }

        public bool Show {
            get { return _show; }
            set { _notifier.SetProperty(ref _show, value); }
        }

        public bool Horizontal {
            get { return _horizontal; }
            set { _notifier.SetProperty(ref _horizontal, value); }
        }

        public int Layers {
            get { return _layers; }
            set { _notifier.SetProperty(ref _layers, value); }
        }

        /// <summary> inside padding </summary>
        public BoundDouble Padding {
            get {
                if (_padding == null)
                    RecalcPadding(null);
                return _padding.Value;
            }
            set {
                this.CheckPadding(value);
                _notifier.SetProperty(ref _padding, value);
            }
        }
        private void RecalcPadding(SizeDouble? old) {
            SizeDouble size = Size;
            var paddingNew = (_padding == null)
                  ? new BoundDouble(size.Width / 2,
                                    size.Height / 2,
                                    _generalModel.Padding.Right,
                                    _generalModel.Padding.Bottom)
                  : this.RecalcPadding(_padding.Value, size, old.Value);
            _notifier.SetProperty(ref _padding, paddingNew, nameof(this.Padding));
        }

        internal struct LineInfo {
            public Color clr;
            public double penWidht;
            public PointDouble from; // start coord
            public PointDouble to;   // end   coord
        }

        /// <summary> get paint information of drawing burger menu model image </summary>
        internal IEnumerable<LineInfo> Coords {
            get {
                if (!Show)
                    return Enumerable.Empty<LineInfo>();

                bool horizontal = Horizontal;
                int layers = Layers;
                var pad = Padding;
                var rc = new RectDouble(pad.Left,
                                        pad.Top,
                                        Size.Width - pad.LeftAndRight,
                                        Size.Height - pad.TopAndBottom); ;
                double penWidth = Math.Max(1, (horizontal ? rc.Height : rc.Width) / (2.0 * layers));
                double rotateAngle = _generalModel.RotateAngle;
                double stepAngle = 360.0 / layers;

                return Enumerable.Range(0, layers)
                   .Select(layerNum => {
                       double layerAlignmentAngle = AnimatedImageModel.FixAngle(layerNum * stepAngle + rotateAngle);
                       double offsetTop = !horizontal ? 0 : layerAlignmentAngle * rc.Height / 360;
                       double offsetLeft = horizontal ? 0 : layerAlignmentAngle * rc.Width / 360;
                       var start = new PointDouble(rc.Left() + offsetLeft,
                                                   rc.Top() + offsetTop);
                       var end = new PointDouble((horizontal ? rc.Right() : rc.Left()) + offsetLeft,
                                                 (horizontal ? rc.Top() : rc.Bottom()) + offsetTop);

                       var hsv = new HSV(Color.Gray);
                       hsv.v *= Math.Sin(layerNum * stepAngle / layers);

                       var li = new LineInfo {
                           clr = hsv.ToColor(),
                           penWidht = penWidth,
                           from = start,
                           to = end
                       };
                       return li;
                   });
            }
        }

        private void OnGeneralModelPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            System.Diagnostics.Debug.Assert(ReferenceEquals(sender, _generalModel));
            if (nameof(IImageModel.Size) == ev.PropertyName) {
                if (ev is PropertyChangedExEventArgs<SizeDouble> evEx)
                    RecalcPadding(evEx.OldValue);
                else
                    throw new Exception();
            }
        }

        public void Dispose() {
            if (_disposed)
                return;
            _disposed = true;

            _generalModel.PropertyChanged -= OnGeneralModelPropertyChanged;
            _notifier.Dispose();
            _generalModel = null;

            GC.SuppressFinalize(this);
        }

    }

}

File: FastMines_core/src/fmg/core/img/BurgerMenuModel.java
package fmg.core.img;

import static fmg.core.img.PropertyConst.PROPERTY_HORIZONTAL;
import static fmg.core.img.PropertyConst.PROPERTY_LAYERS;
import static fmg.core.img.PropertyConst.PROPERTY_PADDING;
import static fmg.core.img.PropertyConst.PROPERTY_ROTATE_ANGLE;
import static fmg.core.img.PropertyConst.PROPERTY_SIZE;

import java.util.function.Consumer;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import fmg.common.Color;
import fmg.common.HSV;
import fmg.common.geom.*;

/** MVC: model of representable menu as horizontal or vertical lines */
public final class BurgerMenuModel implements IImageModel {

    private final SizeDouble size = new SizeDouble(ImageHelper.DEFAULT_IMAGE_SIZE, ImageHelper.DEFAULT_IMAGE_SIZE);
    private final BoundDouble pad = new BoundDouble(ImageHelper.DEFAULT_IMAGE_SIZE / 2.,
                                                    ImageHelper.DEFAULT_IMAGE_SIZE / 2.,
                                                    ImageHelper.DEFAULT_PADDING,
                                                    ImageHelper.DEFAULT_PADDING);
    /** 0° .. +360° */
    private double rotateAngle;
    private boolean horizontal = true;
    private int layers = 3;
    private Consumer<String> changedCallback;

    @Override
    public SizeDouble getSize() {
        return size;
    }

    @Override
    public void setSize(SizeDouble size) {
        if (this.size.equals(size))
            return;

        ImageHelper.checkSize(size);

        SizeDouble oldSize = new SizeDouble(this.size);
        this.size.width  = size.width;
        this.size.height = size.height;

        firePropertyChanged(PROPERTY_SIZE);

        setPadding(ImageHelper.recalcPadding(pad, size, oldSize));
    }

    @Override
    public BoundDouble getPadding() {
        return pad;
    }

    @Override
    public void setPadding(BoundDouble padding) {
        if (pad.equals(padding))
            return;

        ImageHelper.checkPadding(size, padding);

        this.pad.left   = padding.left;
        this.pad.right  = padding.right;
        this.pad.top    = padding.top;
        this.pad.bottom = padding.bottom;

        firePropertyChanged(PROPERTY_PADDING);
    }

    public boolean isHorizontal() { return horizontal; }
    public void   setHorizontal(boolean value) {
        this.horizontal = value;

        firePropertyChanged(PROPERTY_HORIZONTAL);
    }

    public int  getLayers() { return layers; }
    public void setLayers(int value) {
        this.layers = value;

        firePropertyChanged(PROPERTY_LAYERS);
    }

    /** 0° .. +360° */
    public double getRotateAngle() { return rotateAngle; }
    public void setRotateAngle(double value) {
        var old = this.rotateAngle;
        this.rotateAngle = ImageHelper.fixAngle(value);

        if (!DoubleExt.almostEquals(old, this.rotateAngle))
            firePropertyChanged(PROPERTY_ROTATE_ANGLE);
    }

    @Override
    public void setListener(Consumer<String> callback) {
        if ((callback != null) && (changedCallback != null))
            throw new IllegalArgumentException("Can only set the controller once");
        changedCallback = callback;
    }

    public static class LineInfo {
        public Color clr;
        public double penWidht;
        public PointDouble from; // start coord
        public PointDouble to;   // end   coord
    }

    /** get paint information of drawing burger menu model image */
    public Stream<LineInfo> getCoords() {
        RectDouble rc = new RectDouble(pad.left,
                                       pad.top,
                                       getSize().width  - pad.getLeftAndRight(),
                                        getSize().height - pad.getTopAndBottom());
        double penWidth = Math.max(1, (horizontal ? rc.height : rc.width) / (2.0 * layers));
        double stepAngle = 360.0 / layers;

        return IntStream.range(0, layers)
            .mapToObj(layerNum -> {
                double layerAlignmentAngle = ImageHelper.fixAngle(layerNum*stepAngle + rotateAngle);
                double offsetTop  = !horizontal ? 0 : layerAlignmentAngle*rc.height/360;
                double offsetLeft =  horizontal ? 0 : layerAlignmentAngle*rc.width /360;
                PointDouble start = new PointDouble(rc.left() + offsetLeft,
                                                    rc.top()  + offsetTop);
                PointDouble end   = new PointDouble((horizontal ? rc.right() : rc.left()) + offsetLeft,
                                                    (horizontal ? rc.top() : rc.bottom()) + offsetTop);

                HSV hsv = new HSV(Color.Gray());
                hsv.v *= Math.sin(layerNum*stepAngle / layers);

                LineInfo li = new LineInfo();
                li.clr = hsv.toColor();
                li.penWidht = penWidth;
                li.from = start;
                li.to = end;
                return li;
            });
    }

    private void firePropertyChanged(String propertyName) {
        if (changedCallback != null)
            changedCallback.accept(propertyName);
    }

}
